Added tests for reserved core frequency tuning

josclark42 opened the pull request:
Adding a test case for Reserved Core Frequency Tuning.

josclark42 commented:
@yliu127 This is the official PR for Reserved Core Frequency Tuning. I already received feedback from @klaskosk. Please let me know if either of you have suggestions or questions.

yliu127 commented on tests/cnf/ran/powermanagement/internal/helper/helper.go:
```diff
@@ -77,6 +77,28 @@ func SetPowerModeAndWaitForMcpUpdate(perfProfile *nto.Builder, node nodes.Builde
+		return err
+	}
+
+	// Wait for CPU settings to be applied (work in progress)
```
Does this change trigger server reboot?

josclark42 replied:
I changed this to a "PollUntilContextTimeout"

yliu127 commented on tests/cnf/ran/powermanagement/tests/cpufreq.go:
```diff
@@ -0,0 +1,100 @@
+		desiredIsolatedCoreFreq = performancev2.CPUfrequency(2200002)
+		originalIsolatedCPUFreq performancev2.CPUfrequency
+		originalReservedCPUFreq performancev2.CPUfrequency
+		isolatedCPUNumber       = 2
```
this shouldn't be hardcoded. should take config from the same perf profile to get isolated cpu list and pick one cpu from the list.

josclark42 replied:
As per your suggestion, I changed this to get the CPUSets from the PerfProfile, and select the first CPU from each of the two arrays.

yliu127 commented on tests/cnf/ran/powermanagement/tests/cpufreq.go:
```diff
@@ -0,0 +1,100 @@
+		originalIsolatedCPUFreq performancev2.CPUfrequency
+		originalReservedCPUFreq performancev2.CPUfrequency
+		isolatedCPUNumber       = 2
+		ReservedCPUNumber       = 0
```
this shouldn't be hardcoded.

josclark42 replied:
As per your suggestion, I changed this to get the CPUSets from the PerfProfile, and select the first  CPU from each of the two arrays.

yliu127 commented on tests/cnf/ran/powermanagement/tests/cpufreq.go:
```diff
@@ -0,0 +1,100 @@
+
+			By("Compare current isolated core freq to desired isolated core freq")
+			currIsolatedCoreFreq, err := strconv.Atoi(strings.Trim(consoleOut, "\r\n "))
+			Expect(err).ToNot(HaveOccurred(), "strconv.Atoi Failed")
```
consoleOut should be included in error string to help debugging in case failure happens here. As it indicates the cmd output cannot be converted to a number, so it's important to know what is the exact output,

josclark42 replied:
I updated the error to include the console output.

yliu127 commented on tests/cnf/ran/powermanagement/tests/cpufreq.go:
```diff
@@ -0,0 +1,100 @@
+			Expect(err).ToNot(HaveOccurred(), "Failed to set CPU Freq")
+
+			By("Get modified isolated core frequency")
+			spokeCommand := fmt.Sprintf("cat /sys/devices/system/cpu/cpufreq/policy%v/scaling_max_freq |cat -",
```
why adding "| cat -" to mask cmd failure? It would be better to let the test fail when the actual issue is encountered. 

```
[kni@registry.kni-qe-72 ~]$ cat /sys/devices/system/cpu/cpufreq/policy%v/scaling_max_freq |cat -
cat: /sys/devices/system/cpu/cpufreq/policy%v/scaling_max_freq: No such file or directory
[kni@registry.kni-qe-72 ~]$ echo $?
0
[kni@registry.kni-qe-72 ~]$ cat /sys/devices/system/cpu/cpufreq/policy%v/scaling_max_freq
cat: /sys/devices/system/cpu/cpufreq/policy%v/scaling_max_freq: No such file or directory
[kni@registry.kni-qe-72 ~]$ echo $?
1
```

yliu127 replied:
@klaskosk It will make retry pretty much useless if we append "| cat - ". How often do you encounter this issue? We might need to think of other solution to strip off the timestamp.

yliu127 commented on tests/cnf/ran/powermanagement/tests/cpufreq.go:
```diff
@@ -0,0 +1,100 @@
+			err = helper.SetCPUFreqAndWaitForMcpUpdate(perfProfile, *nodeList[0],
+				&desiredIsolatedCoreFreq, &desiredReservedCoreFreq)
+			Expect(err).ToNot(HaveOccurred(), "Failed to set CPU Freq")
+
```
It may be easier to include the verification steps into the SetCPUFreq helper. Because the scope of the helper, it makes sense that way. It also helps verify the revert succeeded end to end in AfterXX.

josclark42 replied:
I refactored the helper to do the varification.

yliu127 commented on tests/cnf/ran/powermanagement/tests/cpufreq.go:
```diff
@@ -0,0 +1,100 @@
+
+var _ = Describe("Per-core runtime power states tuning", Label(tsparams.LabelCPUFrequency), func() {
+	var (
+		nodeList                []*nodes.Builder
```
should remove. This is only used in BeforeXX.

These vars are useful if you set and retrieve the value in different functions. e.g., Set those values once in BeforeAll, and use them in the test function.

yliu127 replied:
This is actually used in AfterXX and It. So closing this comment.

yliu127 commented on tests/cnf/ran/powermanagement/tests/cpufreq.go:
```diff
@@ -0,0 +1,100 @@
+	var (
+		nodeList                []*nodes.Builder
+		perfProfile             *nto.Builder
+		err                     error
```
is this needed?

josclark42 replied:
Kirsten pointed out that it is needed, though not in the global variables. If := is used when getting the perfProfile, the perfProfile var is only scoped to within the beforeEach which causes problems when this var is used later.

yliu127 replied:
yes thats what I meant as well. It is not needed as a shared/global var.  Thanks for the change.

yliu127 commented on tests/cnf/ran/powermanagement/tests/cpufreq.go:
```diff
@@ -0,0 +1,100 @@
+	performancev2 "github.com/openshift/cluster-node-tuning-operator/pkg/apis/performanceprofile/v2"
+)
+
+var _ = Describe("Per-core runtime power states tuning", Label(tsparams.LabelCPUFrequency), func() {
```
This is exactly the same as existing test suite powersave.go, which indicate we should not create separate file for this new test.

josclark42 replied:
I updated the test description.

klaskosk commented on tests/cnf/ran/powermanagement/tests/cpufreq.go:
```diff
@@ -0,0 +1,100 @@
+	)
+
+	BeforeEach(func() {
+		nodeList, err = nodes.List(raninittools.Spoke1APIClient)
```
We don't, I wrote the tests assuming they just wouldn't be run for non SNO clusters. Probably need to update some to not fail but instead skip if the cluster is not SNO then

yliu127 replied:
@klaskosk yes we need to change that assumption. 

Default is all tests can be run on any type of clusters. And if specific tests are for SNO only, then we would skip them when non-SNO clusters are detected. 

@josclark42 let's keep this one as is.

klaskosk replied:
@yliu127 Okay I'll have to do a review of some of the existing tests to add that in

klaskosk commented on tests/cnf/ran/powermanagement/tests/cpufreq.go:
```diff
@@ -0,0 +1,100 @@
+		ReservedCPUNumber       = 0
+	)
+
+	BeforeEach(func() {
```
The BeforeEach instead of BeforeAll was my suggestion since there's only one test in the file the ordered is a bit redundant but if the BeforeEach is expensive then it'd make sense to use continue on failure. My understanding is there's no reboot in this test though

yliu127 replied:
@klaskosk right, but we do revert the config in AfterXX, so best to do that in AfterAll I think if we will have multiple  test cases in the future. But we can also change this structure when we actually have more tests in the future. @josclark42  feel free to keep Before/AfterEach for this PR if you would prefer that way.

josclark42 commented on tests/cnf/ran/powermanagement/tests/cpufreq.go:
```diff
@@ -0,0 +1,100 @@
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"github.com/openshift-kni/eco-goinfra/pkg/nodes"
+	"github.com/openshift-kni/eco-goinfra/pkg/nto" //nolint:misspell
```
@yliu127 Without this comment lint fails with the errror:
`tests/cnf/ran/powermanagement/tests/cpufreq.go:10:44: `nto` is a misspelling of `not` (misspell)`

josclark42 commented:
@yliu127 @klaskosk I made some major changes which resolved a lot of comments and invalidated others. Notably:

- moved the CPU frequency verification from the test to the helper function.
- Added a polling function to the helper to wait for the change to take effect
- Removed unneeded nodelist variable
- removed nodelist as a parameter for the helper function
-  other small changes as per review recommendations.

klaskosk commented on tests/cnf/ran/powermanagement/internal/helper/helper.go:
```diff
@@ -77,6 +81,98 @@ func SetPowerModeAndWaitForMcpUpdate(perfProfile *nto.Builder, node nodes.Builde
+			currIsolatedCoreFreq, err := strconv.Atoi(strings.TrimSpace(consoleOut))
+
+			if err != nil {
+				return false, errors.New(fmt.Sprintf("string conversion failed failed: %v", err))
```
Can use `fmt.Errorf` with the `%w` verb here instead of `%v`

josclark42 replied:
Done.

klaskosk commented on tests/cnf/ran/powermanagement/internal/helper/helper.go:
```diff
@@ -77,6 +81,98 @@ func SetPowerModeAndWaitForMcpUpdate(perfProfile *nto.Builder, node nodes.Builde
+			consoleOut, err := cluster.ExecCommandOnSNO(raninittools.Spoke1APIClient, 3, spokeCommand)
+
+			if err != nil {
+				return false, errors.New(fmt.Sprintf("Command failed: %v", consoleOut))
```
Can use `fmt.Errorf` with the `%w` verb here instead of `%v`

josclark42 replied:
Done.

klaskosk commented on tests/cnf/ran/powermanagement/internal/helper/helper.go:
```diff
@@ -77,6 +81,98 @@ func SetPowerModeAndWaitForMcpUpdate(perfProfile *nto.Builder, node nodes.Builde
+			consoleOut, err := cluster.ExecCommandOnSNO(raninittools.Spoke1APIClient, 3, spokeCommand)
+
+			if err != nil {
+				return false, errors.New(fmt.Sprintf("command failed: %v", consoleOut))
```
See above comments for all errors

josclark42 replied:
Done.

klaskosk commented on tests/cnf/ran/powermanagement/tests/cpufreq.go:
```diff
@@ -0,0 +1,83 @@
+		perfProfile, err = helper.GetPerformanceProfileWithCPUSet()
+		Expect(err).ToNot(HaveOccurred(), "Failed to get performance profile")
+
+		// Get isolated core ID
```
```suggestion
		By("getting isolated core ID")
```

josclark42 replied:
Done.

klaskosk commented on tests/cnf/ran/powermanagement/tests/cpufreq.go:
```diff
@@ -0,0 +1,83 @@
+		isolatedCPUsList := isolatedCPUSet.List()
+		isolatedCPUNumber := isolatedCPUsList[0]
+
+		// Get reserved core ID
```
```suggestion
		By("getting reserved core ID")
```

josclark42 replied:
Done

klaskosk commented on tests/cnf/ran/powermanagement/tests/cpufreq.go:
```diff
@@ -0,0 +1,83 @@
+
+	AfterEach(func() {
+		By("Reverts the CPU frequencies to the original setting")
+		err := helper.SetCPUFreq(perfProfile,
```
can this be one line?

josclark42 replied:
Done- thanks.

klaskosk commented on tests/cnf/ran/powermanagement/tests/cpufreq.go:
```diff
@@ -0,0 +1,83 @@
+
+		It("tests changing reserved and isolated CPU frequencies", func() {
+			By("patch performance profile to set core frequencies")
+			err := helper.SetCPUFreq(perfProfile,
```
Same as above, can this be one line?

josclark42 replied:
Good idea- thanks. Done.

klaskosk replied:
It looks like this is still multiple lines

josclark42 replied:
Fixed for real this time?

klaskosk commented on tests/cnf/ran/powermanagement/internal/tsparams/consts.go:
```diff
@@ -11,7 +11,9 @@ const (
 	LabelPowerSaveTestCases = "powersave"
-	// PowerSaveTimeout is the timeout being used in powersave tests.
+	// LabelCPUFrequency is the label for frequency tuning test cases.
+	LabelCPUFrequency = "cpu_frequency"
```
```suggestion
	LabelCPUFrequency = "cpu-frequency"
```
I've been using hyphens so this is just for consistency

josclark42 replied:
Thank you done.

klaskosk commented on tests/cnf/ran/powermanagement/internal/helper/helper.go:
```diff
@@ -77,6 +81,98 @@ func SetPowerModeAndWaitForMcpUpdate(perfProfile *nto.Builder, node nodes.Builde
+			}
+
+			if currIsolatedCoreFreq != int(*desiredIsolatedCoreFreq) {
+				return false, errors.New("current Isolated CPU Frequency does not match expected frequency")
```
returning false with a non-nil error fails the `wait.PollUntilContextTimeout` so this might need to be `return false, nil`

klaskosk replied:
@josclark42 In this case it's not about `fmt.Errorf` vs `errors.New`, it's that returning an error from `wait.PollUntilContextTimeout` will prevent it from trying again after the interval. I think in this case you still want to keep checking if the current is equal to the desired so you should return a nil error. Otherwise the function will only run once.

josclark42 replied:
Got it now. Thanks for the explanation! Fixed.

klaskosk commented on tests/cnf/ran/powermanagement/internal/helper/helper.go:
```diff
@@ -77,6 +81,98 @@ func SetPowerModeAndWaitForMcpUpdate(perfProfile *nto.Builder, node nodes.Builde
+			}
+
+			if currReservedFreq != int(*desiredReservedCoreFreq) {
+				return false, errors.New("current Reserved CPU Frequency does not match expected frequency")
```
Same as above about non-nil error returns

josclark42 replied:
@klaskosk  I changed this to a fmt.Errorf. I don't think there's a reason to use errors.New.

josclark42 replied:
Fixed

klaskosk reviewed (changes requested):
A few mostly style comments

josclark42 commented on tests/cnf/ran/powermanagement/tests/cpufreq.go:
```diff
@@ -0,0 +1,100 @@
+			isolatedCPUNumber)
+		consoleOut, err := cluster.ExecCommandOnSNO(raninittools.Spoke1APIClient, 3, spokeCommand)
+		Expect(err).ToNot(HaveOccurred(), "Failed to %s", spokeCommand)
+		freqAsInt, err := strconv.Atoi(strings.Trim(consoleOut, "\r\n"))
```
all strings.Trim have been replaced with strings.TrimSpace

josclark42 commented on tests/cnf/ran/powermanagement/internal/helper/helper.go:
```diff
@@ -77,6 +77,28 @@ func SetPowerModeAndWaitForMcpUpdate(perfProfile *nto.Builder, node nodes.Builde
+		return err
+	}
+
+	// Wait for CPU settings to be applied (work in progress)
```
Changed to PollUntilContextTimeout

josclark42 commented on tests/cnf/ran/powermanagement/tests/cpufreq.go:
```diff
@@ -0,0 +1,100 @@
+	BeforeEach(func() {
+		nodeList, err = nodes.List(raninittools.Spoke1APIClient)
+		Expect(err).ToNot(HaveOccurred(), "Failed to get nodes")
+		Expect(len(nodeList)).To(Equal(1), "Currently only SNO clusters are supported")
```
nodeList is not needed at all for this test so I removed it.

klaskosk commented on tests/cnf/ran/powermanagement/internal/helper/helper.go:
```diff
@@ -77,6 +81,98 @@ func SetPowerModeAndWaitForMcpUpdate(perfProfile *nto.Builder, node nodes.Builde
+			consoleOut, err := cluster.ExecCommandOnSNO(raninittools.Spoke1APIClient, 3, spokeCommand)
+
+			if err != nil {
+				return false, fmt.Errorf("command failed: %q", consoleOut)
```
I'm not familiar with `%q`, is there a reason to use it over `%s` here?

josclark42 replied:
In an earlier version lint was complaining about the %s so I changed to %q which satisfied it. I presume it has something to do with the output containing characters that must be double-quoted? 

Regardless, it now passes with a %s which I've changed to for the sake of consistency.

klaskosk commented on tests/cnf/ran/powermanagement/internal/helper/helper.go:
```diff
@@ -77,6 +81,98 @@ func SetPowerModeAndWaitForMcpUpdate(perfProfile *nto.Builder, node nodes.Builde
+			consoleOut, err := cluster.ExecCommandOnSNO(raninittools.Spoke1APIClient, 3, spokeCommand)
+
+			if err != nil {
+				return false, fmt.Errorf("command failed: %q", consoleOut)
```
Same question as below

josclark42 replied:
In an earlier version lint was complaining about the %s so I changed to %q which satisfied it. I presume it has something to do with the output containing characters that must be double-quoted? 

Regardless, it now passes with a %s which I've changed to for the sake of consistency.

josclark42 commented on tests/cnf/ran/powermanagement/internal/helper/helper.go:
```diff
@@ -77,6 +81,98 @@ func SetPowerModeAndWaitForMcpUpdate(perfProfile *nto.Builder, node nodes.Builde
+				return false, err
+			}
+			// Get current isolated core frequency from spoke cluster and compare to desired frequency
+			consoleOut, err := cluster.ExecCommandOnSNO(raninittools.Spoke1APIClient, 3, spokeCommand)
```
Done- thank you.

josclark42 commented on tests/cnf/ran/powermanagement/tests/cpufreq.go:
```diff
@@ -0,0 +1,81 @@
+			Expect(err).ToNot(HaveOccurred(), "Failed to set CPU Freq")
+		})
+
+		Context("Reserved Core Frequency Tuning Test", func() {
```
Done- thank you.

klaskosk commented on tests/cnf/ran/powermanagement/internal/helper/helper.go:
```diff
@@ -77,6 +81,98 @@ func SetPowerModeAndWaitForMcpUpdate(perfProfile *nto.Builder, node nodes.Builde
+	isolatedCPUsList := isolatedCPUSet.List()
+	isolatedCPUNumber := isolatedCPUsList[0]
+
+	spokeCommand := fmt.Sprintf("cat /sys/devices/system/cpu/cpufreq/policy%v/scaling_max_freq |cat -",
```
After a bit of discussion I think the `| cat -` can be removed from this function since the command gets retried so malformed output occasionally isn't an issue

josclark42 replied:
Removed.

klaskosk commented on tests/cnf/ran/powermanagement/internal/helper/helper.go:
```diff
@@ -77,6 +81,98 @@ func SetPowerModeAndWaitForMcpUpdate(perfProfile *nto.Builder, node nodes.Builde
+	reservedCPUsList := reservedCPUSet.List()
+	reservedCCPUNumber := reservedCPUsList[0]
+
+	spokeCommand = fmt.Sprintf("cat /sys/devices/system/cpu/cpufreq/policy%v/scaling_max_freq |cat -",
```
See above about removing `| cat -` in this function specifically

josclark42 replied:
Removed.

klaskosk commented on tests/cnf/ran/powermanagement/tests/cpufreq.go:
```diff
@@ -0,0 +1,81 @@
+		})
+
+		When("reserved and isolated core frequency is configured via PerformanceProfile", func() {
+
```
nitpick, but this blank line can be removed

josclark42 replied:
Removed- thanks.

klaskosk reviewed (commented):
Some more clarity on the `| cat -` situation in the helper

klaskosk commented on tests/cnf/ran/powermanagement/internal/helper/helper.go:
```diff
@@ -77,6 +81,98 @@ func SetPowerModeAndWaitForMcpUpdate(perfProfile *nto.Builder, node nodes.Builde
+
+	// Wait for Isolated CPU Frequency to be updated.
+	err = wait.PollUntilContextTimeout(
+		context.TODO(), 5*time.Second, 1*time.Minute, true, func(ctx context.Context) (bool, error) {
```
Perhaps consider reducing the interval too if it's a quick update since we may have to wait for a couple intervals if the output is malformed

josclark42 replied:
Reduced interval to 2

josclark42 commented on tests/cnf/ran/powermanagement/internal/helper/helper.go:
```diff
@@ -77,6 +81,98 @@ func SetPowerModeAndWaitForMcpUpdate(perfProfile *nto.Builder, node nodes.Builde
+	reservedCPUsList := reservedCPUSet.List()
+	reservedCCPUNumber := reservedCPUsList[0]
+
+	spokeCommand = fmt.Sprintf("cat /sys/devices/system/cpu/cpufreq/policy%v/scaling_max_freq |cat -",
```
Wrapper created.

josclark42 commented on tests/cnf/ran/powermanagement/tests/cpufreq.go:
```diff
@@ -0,0 +1,81 @@
+			originalIsolatedCPUFreq = performancev2.CPUfrequency(freqAsInt)
+
+			By("getting original reserved core frequency")
+			spokeCommand = fmt.Sprintf("cat /sys/devices/system/cpu/cpufreq/policy%v/scaling_max_freq |cat -",
```
Wrapper created.

yliu127 reviewed (approved):
/lgtm

klaskosk reviewed (approved):
/lgtm
